{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b00a771b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading dataset...\n",
      "('A', 'T', 'G', 'T', 'A', 'T', 'C', 'A', 'C', 'T', 'A', 'T', 'T', 'A', 'T', 'A', 'A', 'C', 'C', 'C', 'C', 'G', 'A', 'G', 'G', 'A', 'G', 'A', 'A', 'C', 'C', 'A', 'T', 'T', 'T', 'T', 'T', 'G', 'C', 'G', 'C', 'T', 'A', 'C', 'C', 'T', 'G', 'G', 'G', 'A', 'C', 'G', 'C', 'G', 'T', 'C', 'C', 'A', 'A', 'A', 'C', 'C', 'T', 'T', 'G', 'A', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P', 'P')\n",
      "loaded 2000 lines in dataset\n",
      "Generator_lang(\n",
      "  (fc1): Linear(in_features=128, out_features=79872, bias=True)\n",
      "  (block): Sequential(\n",
      "    (0): ResBlock(\n",
      "      (res_block): Sequential(\n",
      "        (0): ReLU(inplace=True)\n",
      "        (1): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "        (2): ReLU(inplace=True)\n",
      "        (3): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "      )\n",
      "    )\n",
      "    (1): ResBlock(\n",
      "      (res_block): Sequential(\n",
      "        (0): ReLU(inplace=True)\n",
      "        (1): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "        (2): ReLU(inplace=True)\n",
      "        (3): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "      )\n",
      "    )\n",
      "    (2): ResBlock(\n",
      "      (res_block): Sequential(\n",
      "        (0): ReLU(inplace=True)\n",
      "        (1): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "        (2): ReLU(inplace=True)\n",
      "        (3): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "      )\n",
      "    )\n",
      "    (3): ResBlock(\n",
      "      (res_block): Sequential(\n",
      "        (0): ReLU(inplace=True)\n",
      "        (1): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "        (2): ReLU(inplace=True)\n",
      "        (3): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "      )\n",
      "    )\n",
      "    (4): ResBlock(\n",
      "      (res_block): Sequential(\n",
      "        (0): ReLU(inplace=True)\n",
      "        (1): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "        (2): ReLU(inplace=True)\n",
      "        (3): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "      )\n",
      "    )\n",
      "  )\n",
      "  (conv1): Conv1d(512, 5, kernel_size=(1,), stride=(1,))\n",
      ")\n",
      "Discriminator_lang(\n",
      "  (block): Sequential(\n",
      "    (0): ResBlock(\n",
      "      (res_block): Sequential(\n",
      "        (0): ReLU(inplace=True)\n",
      "        (1): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "        (2): ReLU(inplace=True)\n",
      "        (3): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "      )\n",
      "    )\n",
      "    (1): ResBlock(\n",
      "      (res_block): Sequential(\n",
      "        (0): ReLU(inplace=True)\n",
      "        (1): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "        (2): ReLU(inplace=True)\n",
      "        (3): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "      )\n",
      "    )\n",
      "    (2): ResBlock(\n",
      "      (res_block): Sequential(\n",
      "        (0): ReLU(inplace=True)\n",
      "        (1): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "        (2): ReLU(inplace=True)\n",
      "        (3): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "      )\n",
      "    )\n",
      "    (3): ResBlock(\n",
      "      (res_block): Sequential(\n",
      "        (0): ReLU(inplace=True)\n",
      "        (1): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "        (2): ReLU(inplace=True)\n",
      "        (3): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "      )\n",
      "    )\n",
      "    (4): ResBlock(\n",
      "      (res_block): Sequential(\n",
      "        (0): ReLU(inplace=True)\n",
      "        (1): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "        (2): ReLU(inplace=True)\n",
      "        (3): Conv1d(512, 512, kernel_size=(5,), stride=(1,), padding=(2,))\n",
      "      )\n",
      "    )\n",
      "  )\n",
      "  (conv1d): Conv1d(5, 512, kernel_size=(1,), stride=(1,))\n",
      "  (linear): Linear(in_features=79872, out_features=1, bias=True)\n",
      ")\n",
      "Psipred analyzer not evaluating on test set\n",
      "Val Acc:0\n",
      "Loading most recently saved...\n",
      "[*] Checkpoint 12399 found at ./checkpoint/realProt_50aa/!\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 0\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 1\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 2\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 3\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 4\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 5\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 6\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 7\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 8\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 9\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 10\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 11\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 12\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 13\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 14\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 15\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 16\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 17\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 18\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 19\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 20\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 21\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 22\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 23\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 24\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 25\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 26\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 27\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 28\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 29\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 30\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 31\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 32\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 33\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 34\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 35\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 36\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 37\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 38\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 39\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 40\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 41\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 42\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 43\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 44\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 45\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 46\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 47\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 48\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 49\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 50\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 51\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 52\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 53\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 54\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 55\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 56\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 57\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 58\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 59\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 60\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 61\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 62\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 63\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 64\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 65\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 66\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 67\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 68\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 69\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 70\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 71\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 72\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 73\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 74\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 75\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 76\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 77\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 78\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 79\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 80\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 81\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 82\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 83\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 84\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 85\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 86\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 87\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 88\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 89\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 90\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 91\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 92\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 93\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 94\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 95\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 96\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 97\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 98\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 99\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 100\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 101\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 102\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 103\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 104\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 105\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 106\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 107\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 108\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 109\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 110\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 111\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 112\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 113\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 114\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 115\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 116\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 117\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 118\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 119\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 120\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 121\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 122\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 123\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 124\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 125\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 126\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 127\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 128\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 129\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 130\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 131\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 132\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 133\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 134\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 135\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 136\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 137\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 138\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 139\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 140\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 141\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 142\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 143\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 144\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 145\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 146\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 147\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 148\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 149\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 150\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 151\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 152\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 153\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 154\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 155\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 156\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 157\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 158\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 159\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 160\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 161\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 162\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 163\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 164\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 165\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 166\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 167\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 168\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 169\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 170\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 171\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 172\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 173\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 174\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 175\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 176\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 177\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 178\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 179\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 180\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 181\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 182\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 183\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 184\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 185\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 186\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 187\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 188\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 189\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 190\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 191\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 192\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 193\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 194\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 195\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 196\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 197\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 198\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 199\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 200\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 201\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 202\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 203\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 204\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 205\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 206\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 207\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 208\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 209\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 210\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 211\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 212\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 213\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 214\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 215\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 216\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 217\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 218\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 219\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 220\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 221\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 222\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 223\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 224\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 225\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 226\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 227\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 228\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 229\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 230\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 231\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 232\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 233\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 234\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 235\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 236\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 237\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 238\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 239\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 240\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 241\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 242\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 243\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 244\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 245\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 246\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 247\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 248\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 249\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 250\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 251\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 252\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 253\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 254\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 255\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 256\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 257\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 258\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 259\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 260\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 261\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 262\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 263\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 264\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 265\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 266\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 267\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 268\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 269\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 270\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 271\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 272\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 273\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 274\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 275\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 276\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 277\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 278\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 279\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 280\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 281\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 282\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 283\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 284\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 285\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 286\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 287\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 288\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 289\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 290\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 291\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 292\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 293\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 294\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 295\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 296\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 297\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 298\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 299\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 300\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 301\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 302\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 303\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 304\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 305\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 306\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 307\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 308\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 309\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 310\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 311\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 312\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 313\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 314\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 315\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 316\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 317\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 318\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 319\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 320\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 321\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 322\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 323\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 324\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 325\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 326\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 327\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 328\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 329\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 330\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 331\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 332\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 333\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 334\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 335\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 336\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 337\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 338\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 339\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 340\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 341\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 342\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 343\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 344\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 345\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 346\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 347\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 348\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 349\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 350\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 351\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 352\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 353\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 354\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 355\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 356\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 357\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 358\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 359\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 360\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 361\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 362\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 363\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 364\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 365\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 366\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 367\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 368\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 369\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 370\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 371\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 372\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 373\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 374\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 375\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 376\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 377\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 378\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 379\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 380\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 381\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 382\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 383\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 384\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 385\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 386\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 387\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 388\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 389\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 390\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 391\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 392\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 393\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 394\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 395\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 396\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 397\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 398\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 399\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 400\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 401\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 402\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 403\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 404\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 405\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 406\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 407\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 408\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 409\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 410\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 411\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 412\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 413\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 414\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 415\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 416\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 417\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 418\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 419\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 420\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 421\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 422\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 423\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 424\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 425\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 426\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 427\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 428\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 429\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 430\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 431\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 432\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 433\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 434\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 435\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 436\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 437\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 438\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 439\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 440\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 441\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 442\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 443\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 444\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 445\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 446\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 447\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 448\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 449\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 450\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 451\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 452\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 453\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 454\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 455\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 456\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 457\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 458\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 459\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 460\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 461\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 462\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 463\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 464\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 465\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 466\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 467\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 468\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 469\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 470\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 471\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 472\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 473\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 474\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 475\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 476\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 477\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 478\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 479\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 480\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 481\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 482\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 483\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 484\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 485\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 486\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 487\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 488\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 489\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 490\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 491\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 492\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 493\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 494\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 495\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 496\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 497\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 498\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 499\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 500\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 501\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 502\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 503\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 504\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 505\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 506\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 507\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 508\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 509\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 510\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 511\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 512\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 513\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 514\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 515\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 516\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 517\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 518\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 519\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 520\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 521\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 522\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 523\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 524\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 525\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 526\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 527\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 528\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 529\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 530\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 531\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 532\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 533\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 534\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 535\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 536\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 537\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 538\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 539\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 540\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 541\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 542\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 543\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 544\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 545\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 546\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 547\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 548\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 549\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 550\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 551\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 552\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 553\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 554\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 555\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 556\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 557\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 558\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 559\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 560\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 561\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 562\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 563\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 564\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 565\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 566\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 567\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 568\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 569\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 570\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 571\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 572\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 573\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 574\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 575\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 576\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 577\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 578\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 579\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 580\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 581\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 582\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 583\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 584\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 585\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 586\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 587\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 588\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 589\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 590\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 591\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 592\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 593\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 594\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 595\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 596\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 597\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 598\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 599\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 600\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 601\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 602\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 603\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 604\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 605\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 606\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 607\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 608\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 609\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 610\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 611\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 612\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 613\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 614\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 615\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 616\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 617\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 618\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 619\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 620\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 621\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 622\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 623\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 624\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 625\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 626\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 627\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 628\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 629\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 630\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 631\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 632\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 633\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 634\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 635\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 636\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 637\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 638\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 639\n",
      "Adding 0 positive sequences \n",
      " Cutoff 5\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Iteration 9 - loss_d: [-7.6323357], loss_g: -0.021928519010543823, w_dist: [8.774441], grad_penalty: 1.1421051025390625\n",
      "Iteration 19 - loss_d: [-5.078023], loss_g: 1.4229832887649536, w_dist: [5.6469173], grad_penalty: 0.5688945055007935\n",
      "Iteration 29 - loss_d: [-5.617543], loss_g: 2.915849208831787, w_dist: [7.3135214], grad_penalty: 1.695978045463562\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 0\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 1\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 2\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 3\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 5\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 6\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 7\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 8\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 9\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 10\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 11\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 12\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 13\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 14\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 15\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 16\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 17\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 19\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 20\n",
      "Error! Protein too short.\n",
      "Error in sequence 21\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 22\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 23\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 24\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 26\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 27\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 28\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 29\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 30\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 31\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 32\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 33\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 34\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 35\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 36\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 37\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 38\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 39\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 40\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 41\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 42\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 43\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 44\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 45\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 46\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 47\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 48\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 49\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 50\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 52\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 53\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 54\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 56\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 57\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 58\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 59\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 61\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 62\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 64\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 65\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 66\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 67\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 68\n",
      "Error! Invalid Codon PPP in ATGACTAAGTCCTGGGACGGTTACCACGTGTGGTGCGCGCACAGAPPPPPPPPPPAAPPAPAGTCAGGCCACGAAGGCAACAAACAGTGACTAACGGAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 69\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 70\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 71\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 72\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 73\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 74\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 75\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 76\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 77\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 78\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 79\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 80\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 81\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 83\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 84\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 85\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 86\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 87\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 88\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 89\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 90\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 91\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 92\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 94\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 95\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 96\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 97\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 98\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 99\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 100\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 101\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 102\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 103\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 104\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 105\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 106\n",
      "Error! Invalid Codon AAP in ATGATGCTCCACCGTTCTGATAAAAACTACTTCTCTAACTTCATGACCTATAAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPGAAACCAGAATGCGTAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 107\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 108\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 109\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 111\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 112\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 113\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 114\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 115\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 116\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 117\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 118\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 119\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 120\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 121\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 122\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 123\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 124\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 125\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 127\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 128\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 129\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 130\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 131\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 132\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 133\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 134\n",
      "Error! Protein too short.\n",
      "Error in sequence 135\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 136\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 137\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 138\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 139\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 140\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 141\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 142\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 143\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 144\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 145\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 146\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 147\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 148\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 149\n",
      "Error! Invalid Codon GAP in ATGATGCCTTGGGTGAAGTTGGAACAGTACCTATTGGCCATCGAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 150\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 151\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 152\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 153\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 154\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 155\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 156\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 157\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 158\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 159\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 160\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 161\n",
      "Error! Invalid Codon PPP in ATGATCCGGACTTCCTTCTCTAAACTGAATCACTTAGTTGCTCGCGAGACAACTCTCCTACATTCCGACGAGAAAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 162\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 163\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 164\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 165\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 166\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 167\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 168\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 169\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 172\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 173\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 174\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 177\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 178\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 179\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 180\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 181\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 182\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 183\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 184\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 185\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 186\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 187\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 188\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 189\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 190\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 191\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 192\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 193\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 194\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 195\n",
      "Error! Invalid Codon PPP in ATGACTATGTCGGATATTCCAATAGTTAATCCGATGTGGGGACACCCAGACGAAATCATAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPAAGAAGCGCTCAATPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 196\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 198\n",
      "Error! Invalid Codon PPT in ATGACGGTTCTGATGACAATTGCTAGACTGAGTCATCTTGTAPPTGGCAAAACAACTCTGGACTAGPAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 199\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 200\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 201\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 202\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 203\n",
      "Error! Invalid Codon PPP in ATGATGTATCTGCTGGCCACAAATAACTGTGGGAAGCAAGCTCTAACGGATCCGGCGCTGGGAGGGCCCGTGTTGGTGAATGAAGTCACGGAATTTCAATTAATTGCGCGACTTGTGAAGCACTTGAATTGTATGGGAPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 204\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 205\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 206\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 208\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 209\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 212\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 213\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 215\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 216\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 217\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 218\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 219\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 220\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 221\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 222\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 223\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 224\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 225\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 226\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 227\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 228\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 229\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 230\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 231\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 232\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 233\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 234\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 236\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 237\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 238\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 239\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 240\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 241\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 242\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 243\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 244\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 245\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 246\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 248\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 249\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 250\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 251\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 252\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 253\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 254\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 255\n",
      "Error! Invalid Codon PPP in ATGATTGATTCAGCTCCAGCTAACTTAGATTCATTGATTGAAACAGGCATTTTTGATTTAGCGTCGAAAATCATCGCTAAGAGTAATPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 256\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 257\n",
      "Error! Invalid Codon PAG in ATGACTGTGAACATGCCTCCGGTAGGTGAGATGTTAATAGACATAGGAATTGTTGGGACAAACAGGAATGGTGAAPAGCGAAAAAAGACAGTGAATCCAGTTCTCCAGACTAATACGACTGCCACGGPAPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 258\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 259\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 260\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 261\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 262\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 263\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 264\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 265\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 266\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 267\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 268\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 269\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 270\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 271\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 272\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 273\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 274\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 275\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 277\n",
      "Error! Invalid Codon GAP in ATGCTGATCAAACACTGTGATTTAACTGATCAAATTGCCAATGCATGTTGCGTTGAATTGAAGCACAAGGGAGGAAACCCATCCAAAGGGGAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 278\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 279\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 281\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 282\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 283\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 284\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 285\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 287\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 288\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 289\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 290\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 291\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 292\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 293\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 294\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 295\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 296\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 297\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 298\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 299\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 300\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 301\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 302\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 303\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 304\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 305\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 306\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 307\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 308\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 309\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 311\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 312\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 313\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 314\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 315\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 316\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 318\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 319\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 320\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 321\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 322\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 323\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 324\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 325\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 326\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 327\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 328\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 329\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 330\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 331\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 332\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 333\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 334\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 335\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 336\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 337\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 338\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 340\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 341\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 342\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 343\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 344\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 345\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 346\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 347\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 348\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 349\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 350\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 351\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 352\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 353\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 354\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 355\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 356\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 357\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 358\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 359\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 360\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 361\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 362\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 363\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 364\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 365\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 366\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 367\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 368\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 369\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 370\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 371\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 372\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 373\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 374\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 375\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 376\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 377\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 378\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 379\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 380\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 381\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 382\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 383\n",
      "Error! Invalid Codon PPP in ATGATATTTGACGTTAGGTTACGACAAAAGCTTGCTGCGCATAAGGTAPPPPPPPPPPPPPPPPPPPPPPPPPACAATGATAGAAGAAAAAGTCGGGTTTGATTCTCCAAACGGATGCTTCTATATCGTCCGAPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 384\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 385\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 386\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 387\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 388\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 389\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 391\n",
      "Error! Invalid Codon PPP in ATGAAGATTCTGACTACGGGCCTCGCTAATCTCGACCGGGCTGAAATTCCGGTCATTAAGGCTGAGGTGGGGTGGAGGACTGTCGCTGCGATACATGACTTCTGCGAATATAGTGAGGTCAAGGGCCGAPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 392\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 393\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 394\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 395\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 396\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 397\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 399\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 400\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 401\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 403\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 406\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 407\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 408\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 409\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 410\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 411\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 412\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 413\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 414\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 415\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 416\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 417\n",
      "Error! Invalid Codon PPP in ATGCTTCCCAGGCTCCATCCAGACGCGCTGTTGGAAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 418\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 419\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 420\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 421\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 422\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 423\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 424\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 425\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 426\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 427\n",
      "Error! Invalid Codon GAP in ATGACGATTAATACGTTTAGAGTGACGCAGGTGAATATTGCTATGAATTATGAPPPPPPPPPPPPPPPPPATCACAATPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 428\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 429\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 430\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 432\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 433\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 434\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 435\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 436\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 437\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 438\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 439\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 440\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 441\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 443\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 444\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 445\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 446\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 447\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 448\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 449\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 450\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 451\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 452\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 454\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 455\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 456\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 457\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 458\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 459\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 460\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 461\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 462\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 464\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 465\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 466\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 467\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 468\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 469\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 471\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 472\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 473\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 474\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 475\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 476\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 477\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 478\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 479\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 480\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 481\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 482\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 483\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 484\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 485\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 486\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 487\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 489\n",
      "Error! Invalid Codon PAP in ATGATTCACTCCTCTGGAATTACCCAGTGTCTTGAACGAAATATGCAACTCCATGGCAAAGAAPAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 490\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 491\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 492\n",
      "Error! Invalid Codon GAP in ATGACGCAGGTGTGCGTGAGGAAAAAGGGTGCCTATGCTAAAGTTGCACTAGCTGGACTGCCACCAAATCCGCATCCGCCTTTAAATGGCGTCACGTGCCGCCACGGTCTCCCTGTTTCTGTCGGCAGCAGCGAPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 493\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 494\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 495\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 496\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 497\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 498\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 499\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 500\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 501\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 502\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 503\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 504\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 505\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 506\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 507\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 508\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 509\n",
      "Error! Invalid Codon AGP in ATGATTGCCTATGCCATAGTGCTCACGCCTTATGTTAAAGTCCTATCAATTGATAGGAGCACTGACGAAAAAAACAGPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 510\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 511\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 512\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 513\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 514\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 515\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 516\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 517\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 518\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 519\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 520\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 521\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 522\n",
      "Error! Invalid Codon PPP in ATGATCATGTTGAGCTTAAATGAAATGAAGTCTGTGGACCATAGCCTCAGACGCTATPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 523\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 524\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 525\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 526\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 527\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 528\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 529\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 530\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 532\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 533\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 534\n",
      "Error! Invalid Codon PPP in ATGAATGGAGCCATATTCCCGCGCCAGCTGGTGCCGAACCGTGTGCAACATGTCGCTCTTGTCGATPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 535\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 536\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 537\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 538\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 539\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 540\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 541\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 542\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 543\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 544\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 545\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 546\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 547\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 548\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 549\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 551\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 553\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 554\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 555\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 557\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 558\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 559\n",
      "Error! Invalid Codon PPP in ATGATCACGTGTAGATTTGCCAAGACCTATGGGTTGGTAAACACTGATCGAATTCCTGACCATGAAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 560\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 562\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 563\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 564\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 565\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 566\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 567\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 568\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 569\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 570\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 572\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 573\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 574\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 575\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 576\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 577\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 578\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 580\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 581\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 582\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 583\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 584\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 585\n",
      "Error! Invalid Codon GAP in ATGAGGATTCCCACCGCTGCGGGGATGGCTCTAAATGGCACTAAGAATTTCAGAGTGTTGCCGAATCACCGCTTTCCTGTTCTTGAACGATTGACTATATGTCTAGATGTTGCGGAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 586\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 587\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 588\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 589\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 590\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 591\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 592\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 593\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 594\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 595\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 596\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 597\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 598\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 599\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 600\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 601\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 602\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 603\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 604\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 605\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 606\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 607\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 608\n",
      "Error! Invalid Codon PPP in ATGATTGCTAATGAAGGCTACTTTAGGGTACATGATGATPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 609\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 610\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 611\n",
      "Error! Invalid Codon GGP in ATGAACATTACTCTCTCCACCGATTACATGACTGTGATGTTGGGPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 612\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 613\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 614\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 615\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 616\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 617\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 618\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 619\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 620\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 621\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 622\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 623\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 624\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 625\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 626\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 627\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 628\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 629\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 630\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 631\n",
      "Error! Invalid Codon PPP in ATGCCTCGTGCTCGGTTATACGAACTGGACGGCTTCTTGAATACATACATGACAATGGGAGGTATCCGCGCGTCATGTCGAATGAGATGCGCATGTTGCGATGAAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 632\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 633\n",
      "Error! Invalid Codon GPA in ATGACTGCGAATTTGACAATCCTCTGGAGGAAAAGGAATTGGATTCGCCATTATGTGCAAAATAAGGCTAGCGPAPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP\n",
      "Error! No valid stop codon.\n",
      "Error in sequence 634\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 635\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 636\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 638\n",
      "Not valid gene (no ATG)\n",
      "Error in sequence 639\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/lib/python3.9/multiprocessing/pool.py:853\u001b[0m, in \u001b[0;36mIMapIterator.next\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    852\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 853\u001b[0m     item \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_items\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpopleft\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    854\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mIndexError\u001b[39;00m:\n",
      "\u001b[0;31mIndexError\u001b[0m: pop from an empty deque",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[0;32mIn [6]\u001b[0m, in \u001b[0;36m<cell line: 252>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    250\u001b[0m     model\u001b[38;5;241m.\u001b[39mtrain_model(args\u001b[38;5;241m.\u001b[39mload_dir)\n\u001b[1;32m    252\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;18m__name__\u001b[39m \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m__main__\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[0;32m--> 253\u001b[0m     \u001b[43mmain\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "Input \u001b[0;32mIn [6]\u001b[0m, in \u001b[0;36mmain\u001b[0;34m()\u001b[0m\n\u001b[1;32m    248\u001b[0m args \u001b[38;5;241m=\u001b[39m parser\u001b[38;5;241m.\u001b[39mparse_known_args()[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m    249\u001b[0m model \u001b[38;5;241m=\u001b[39m WGAN_LangGP(run_name\u001b[38;5;241m=\u001b[39margs\u001b[38;5;241m.\u001b[39mrun_name)\n\u001b[0;32m--> 250\u001b[0m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtrain_model\u001b[49m\u001b[43m(\u001b[49m\u001b[43margs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mload_dir\u001b[49m\u001b[43m)\u001b[49m\n",
      "Input \u001b[0;32mIn [6]\u001b[0m, in \u001b[0;36mWGAN_LangGP.train_model\u001b[0;34m(self, load_dir)\u001b[0m\n\u001b[1;32m    147\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m epoch \u001b[38;5;241m%\u001b[39m \u001b[38;5;241m2\u001b[39m \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m: \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msave_model(epoch)\n\u001b[1;32m    148\u001b[0m sampled_seqs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msample(num_batches_sample, epoch)\n\u001b[0;32m--> 149\u001b[0m preds \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43manalyzer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpredict_model\u001b[49m\u001b[43m(\u001b[49m\u001b[43msampled_seqs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    150\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m min_helix_len \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    151\u001b[0m     min_helix_len \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mmax(preds)\n",
      "File \u001b[0;32m~/fbgan-master/psipred_wrapper.py:59\u001b[0m, in \u001b[0;36mPsipredAnalyzer.predict_model\u001b[0;34m(self, input_seqs)\u001b[0m\n\u001b[1;32m     57\u001b[0m     out_files \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m [out_file]\n\u001b[1;32m     58\u001b[0m pool \u001b[38;5;241m=\u001b[39m Pool(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnum_cpu)\n\u001b[0;32m---> 59\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, returncode \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(pool\u001b[38;5;241m.\u001b[39mimap(partial(call, shell\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m), commands)):\n\u001b[1;32m     60\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m returncode \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m     61\u001b[0m         \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[38;5;124m command failed: \u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m (i, returncode))\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.9/multiprocessing/pool.py:858\u001b[0m, in \u001b[0;36mIMapIterator.next\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    856\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_pool \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    857\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mStopIteration\u001b[39;00m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28mNone\u001b[39m\n\u001b[0;32m--> 858\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_cond\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    859\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    860\u001b[0m     item \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_items\u001b[38;5;241m.\u001b[39mpopleft()\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.9/threading.py:312\u001b[0m, in \u001b[0;36mCondition.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    310\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:    \u001b[38;5;66;03m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[39;00m\n\u001b[1;32m    311\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 312\u001b[0m         \u001b[43mwaiter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43macquire\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    313\u001b[0m         gotit \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    314\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#    Copyright (C) 2018 Anvita Gupta\n",
    "#\n",
    "#    This program is free software: you can redistribute it and/or  modify\n",
    "#    it under the terms of the GNU Affero General Public License, version 3,\n",
    "#    as published by the Free Software Foundation.\n",
    "#\n",
    "#    This program is distributed in the hope that it will be useful,\n",
    "#    but WITHOUT ANY WARRANTY; without even the implied warranty of\n",
    "#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n",
    "#    GNU Affero General Public License for more details.\n",
    "#\n",
    "#    You should have received a copy of the GNU Affero General Public License\n",
    "#    along with this program.  If not, see <http://www.gnu.org/licenses/>.\n",
    "#\n",
    "\n",
    "import torch\n",
    "from torch import optim\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "import torch.autograd as autograd\n",
    "from torch.autograd import Variable\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "import os, math, glob, argparse\n",
    "from utils.torch_utils import *\n",
    "from utils.utils import *\n",
    "from psipred_wrapper import *\n",
    "import matplotlib.pyplot as plt\n",
    "import utils.language_helpers as language_helpers\n",
    "plt.switch_backend('agg')\n",
    "import numpy as np\n",
    "from models import *\n",
    "\n",
    "class WGAN_LangGP():\n",
    "    def __init__(self, batch_size=64, lr=0.0001, num_epochs=150, seq_len = 156, data_dir='./data/dna_uniprot_under_50_reviewed.fasta', \\\n",
    "        run_name='test', hidden=512, d_steps = 10, max_examples=2000):\n",
    "        self.hidden = hidden\n",
    "        self.batch_size = batch_size\n",
    "        self.lr = lr\n",
    "        self.n_epochs = num_epochs\n",
    "        self.seq_len = seq_len\n",
    "        self.d_steps = d_steps\n",
    "        self.g_steps = 1\n",
    "        self.lamda = 10 #lambda\n",
    "        self.run_name = run_name\n",
    "        self.checkpoint_dir = './checkpoint/' + run_name + \"/\"\n",
    "        self.sample_dir = './samples/' + run_name + \"/\"\n",
    "        self.load_data(data_dir, max_examples)\n",
    "        if not os.path.exists(self.checkpoint_dir): os.makedirs(self.checkpoint_dir)\n",
    "        if not os.path.exists(self.sample_dir): os.makedirs(self.sample_dir)\n",
    "        self.use_cuda = True if torch.cuda.is_available() else False\n",
    "        self.build_model()\n",
    "\n",
    "    def build_model(self):\n",
    "        self.G = Generator_lang(len(self.charmap), self.seq_len, self.batch_size, self.hidden)\n",
    "        self.D = Discriminator_lang(len(self.charmap), self.seq_len, self.batch_size, self.hidden)\n",
    "        if self.use_cuda:\n",
    "            self.G.cuda()\n",
    "            self.D.cuda()\n",
    "        print(self.G)\n",
    "        print(self.D)\n",
    "        self.G_optimizer = optim.Adam(self.G.parameters(), lr=self.lr, betas=(0.5, 0.9))\n",
    "        self.D_optimizer = optim.Adam(self.D.parameters(), lr=self.lr, betas=(0.5, 0.9))\n",
    "        self.analyzer = PsipredAnalyzer(self.run_name) #from PYTORCH\n",
    "        val_loss, val_acc = self.analyzer.evaluate_model()\n",
    "        print(\"Val Acc:{}\".format(val_acc))\n",
    "\n",
    "    def load_data(self, datadir, max_examples=1e6):\n",
    "        self.data, self.charmap, self.inv_charmap = language_helpers.load_dataset(\n",
    "            max_length=self.seq_len,\n",
    "            max_n_examples=max_examples,\n",
    "            data_dir=datadir\n",
    "        )\n",
    "        self.labels = np.zeros(len(self.data)) #this marks at which epoch this data was added\n",
    "\n",
    "    def remove_old_indices(self, numToAdd):\n",
    "        toRemove = np.argsort(self.labels)[:numToAdd]\n",
    "        self.data = [d for i,d in enumerate(self.data) if i not in toRemove]\n",
    "        self.labels = np.delete(self.labels, toRemove)\n",
    "\n",
    "    def save_model(self, epoch):\n",
    "        torch.save(self.G.state_dict(), self.checkpoint_dir + \"G_weights_{}.pth\".format(epoch))\n",
    "        torch.save(self.D.state_dict(), self.checkpoint_dir + \"D_weights_{}.pth\".format(epoch))\n",
    "\n",
    "    def load_model(self, directory = '', iteration=None):\n",
    "        '''\n",
    "            Load model parameters from most recent epoch\n",
    "        '''\n",
    "        if len(directory) == 0:\n",
    "            directory = self.checkpoint_dir\n",
    "        list_G = glob.glob(directory + \"G*.pth\")\n",
    "        list_D = glob.glob(directory + \"D*.pth\")\n",
    "        if len(list_G) == 0:\n",
    "            print(\"[*] Checkpoint not found! Starting from scratch.\")\n",
    "            return 1 #file is not there\n",
    "        if iteration is None:\n",
    "            print(\"Loading most recently saved...\")\n",
    "            G_file = max(list_G, key=os.path.getctime)\n",
    "            D_file = max(list_D, key=os.path.getctime)\n",
    "        else:\n",
    "            G_file = \"G_weights_{}.pth\".format(iteration)\n",
    "            D_file = \"D_weights_{}.pth\".format(iteration)\n",
    "        epoch_found = int( (G_file.split('_')[-1]).split('.')[0])\n",
    "        print(\"[*] Checkpoint {} found at {}!\".format(epoch_found, directory))\n",
    "        self.G.load_state_dict(torch.load(G_file, map_location='cpu'))\n",
    "        self.D.load_state_dict(torch.load(D_file, map_location='cpu'))\n",
    "        return epoch_found\n",
    "\n",
    "    def calc_gradient_penalty(self, real_data, fake_data):\n",
    "        alpha = torch.rand(self.batch_size, 1, 1)\n",
    "        alpha = alpha.view(-1,1,1)\n",
    "        alpha = alpha.expand_as(real_data)\n",
    "        alpha = alpha.cuda() if self.use_cuda else alpha\n",
    "        interpolates = alpha * real_data + ((1 - alpha) * fake_data)\n",
    "\n",
    "        interpolates = interpolates.cuda() if self.use_cuda else interpolates\n",
    "        interpolates = autograd.Variable(interpolates, requires_grad=True)\n",
    "\n",
    "        disc_interpolates = self.D(interpolates)\n",
    "\n",
    "        gradients = autograd.grad(outputs=disc_interpolates, inputs=interpolates,\n",
    "                                  grad_outputs=torch.ones(disc_interpolates.size()).cuda() \\\n",
    "                                  if self.use_cuda else torch.ones(disc_interpolates.size()),\n",
    "                                  create_graph=True, retain_graph=True, only_inputs=True)[0]\n",
    "\n",
    "        gradient_penalty = ((gradients.norm(2, dim=1).norm(2,dim=1) - 1) ** 2).mean() * self.lamda\n",
    "        return gradient_penalty\n",
    "\n",
    "    def train_model(self, load_dir):\n",
    "        self.load_model(load_dir)\n",
    "        losses_f = open(self.checkpoint_dir + \"losses.txt\",'a+')\n",
    "        d_fake_losses, d_real_losses, grad_penalties = [],[],[]\n",
    "        G_losses, D_losses, W_dist = [],[],[]\n",
    "\n",
    "        one = torch.FloatTensor([1])\n",
    "        one = one.cuda() if self.use_cuda else one\n",
    "        one_neg = one * -1\n",
    "\n",
    "        table = np.arange(len(self.charmap)).reshape(-1, 1)\n",
    "        one_hot = OneHotEncoder()\n",
    "        one_hot.fit(table)\n",
    "        num_batches_sample = 10\n",
    "        n_batches = int(len(self.data)/self.batch_size)\n",
    "        i = 0\n",
    "        min_helix_len = 5\n",
    "        alpha = 0\n",
    "        for epoch in range(1, self.n_epochs+1):\n",
    "            if epoch % 2 == 0: self.save_model(epoch)\n",
    "            sampled_seqs = self.sample(num_batches_sample, epoch)\n",
    "            preds = self.analyzer.predict_model(sampled_seqs)\n",
    "            if min_helix_len == 0:\n",
    "                min_helix_len = np.max(preds)\n",
    "                norm_preds = preds/min_helix_len\n",
    "                good_indices = (norm_preds > 0.5).nonzero()[0]\n",
    "            else:\n",
    "                good_indices = (preds > min_helix_len).nonzero()[0]\n",
    "            with open(self.sample_dir + \"sampled_{}_preds.txt\".format(epoch), 'w+') as f:\n",
    "                f.writelines(['Cuttoff:{}\\n'.format(min_helix_len)])\n",
    "                f.writelines([s + '\\t' + str(preds[j]) + '\\n' for j, s in enumerate(sampled_seqs)])\n",
    "            pos_seqs = [list(sampled_seqs[i]) for i in good_indices]\n",
    "            print(\"Adding {} positive sequences \\n Cutoff {}\".format(len(pos_seqs), min_helix_len))\n",
    "            with open(self.checkpoint_dir + \"positives.txt\",'a+') as f:\n",
    "                f.write(\"Epoch: {} \\t Pos: {}\\n\".format(epoch, len(pos_seqs)/float(len(sampled_seqs))))\n",
    "            self.remove_old_indices(len(pos_seqs))\n",
    "            self.data += pos_seqs\n",
    "            self.labels = np.concatenate([self.labels, np.repeat(epoch, len(pos_seqs))] )\n",
    "            perm = np.random.permutation(len(self.data))\n",
    "            self.data = [self.data[i] for i in perm]\n",
    "            self.labels = self.labels[perm]\n",
    "\n",
    "            for idx in range(n_batches):\n",
    "                _data = np.array(\n",
    "                    [[self.charmap[c] for c in l] for l in self.data[idx*self.batch_size:(idx+1)*self.batch_size]],\n",
    "                    dtype='int32'\n",
    "                )\n",
    "                data_one_hot = one_hot.transform(_data.reshape(-1, 1)).toarray().reshape(self.batch_size, -1, len(self.charmap))\n",
    "                real_data = torch.Tensor(data_one_hot)\n",
    "                real_data = to_var(real_data)\n",
    "                for p in self.D.parameters():  # reset requires_grad\n",
    "                    p.requires_grad = True  # they are set to False below in netG update\n",
    "                for _ in range(self.d_steps): # Train D\n",
    "                    self.D.zero_grad()\n",
    "                    d_real_pred = self.D(real_data)\n",
    "                    d_real_err = torch.mean(d_real_pred) #want to push d_real as high as possible\n",
    "                    d_real_err = torch.reshape(d_real_err,(-1,))\n",
    "                    d_real_err.backward(one_neg)\n",
    "\n",
    "                    z_input = to_var(torch.randn(self.batch_size, 128))\n",
    "                    d_fake_data = self.G(z_input).detach()\n",
    "                    d_fake_pred = self.D(d_fake_data)\n",
    "                    d_fake_err = torch.mean(d_fake_pred) #want to push d_fake as low as possible\n",
    "                    d_fake_err = torch.reshape(d_fake_err,(-1,))\n",
    "                    d_fake_err.backward(one)\n",
    "\n",
    "                    gradient_penalty = self.calc_gradient_penalty(real_data.data, d_fake_data.data)\n",
    "                    gradient_penalty.backward()\n",
    "\n",
    "                    d_err = d_fake_err - d_real_err + gradient_penalty\n",
    "                    self.D_optimizer.step()\n",
    "\n",
    "                # Append things for logging\n",
    "                d_fake_np, d_real_np, gp_np = (d_fake_err.data).cpu().numpy(), \\\n",
    "                        (d_real_err.data).cpu().numpy(), (gradient_penalty.data).cpu().numpy()\n",
    "                grad_penalties.append(gp_np)\n",
    "                d_real_losses.append(d_real_np)\n",
    "                d_fake_losses.append(d_fake_np)\n",
    "                D_losses.append(d_fake_np - d_real_np + gp_np)\n",
    "                W_dist.append(d_real_np - d_fake_np)\n",
    "                # Train G\n",
    "                for p in self.D.parameters():\n",
    "                    p.requires_grad = False  # to avoid computation\n",
    "\n",
    "                self.G.zero_grad()\n",
    "                z_input = to_var(torch.randn(self.batch_size, 128))\n",
    "                g_fake_data = self.G(z_input)\n",
    "                dg_fake_pred = self.D(g_fake_data)\n",
    "                g_err = -torch.mean(dg_fake_pred)\n",
    "                g_err.backward()\n",
    "                self.G_optimizer.step()\n",
    "                G_losses.append((g_err.data).cpu().numpy())\n",
    "                if i % 10 == 9:\n",
    "                    summary_str = 'Iteration {} - loss_d: {}, loss_g: {}, w_dist: {}, grad_penalty: {}'\\\n",
    "                        .format(i, (d_err.data).cpu().numpy(),\n",
    "                        (g_err.data).cpu().numpy(), ((d_real_err - d_fake_err).data).cpu().numpy(), gp_np)\n",
    "                    print(summary_str)\n",
    "                    losses_f.write(summary_str)\n",
    "                    plot_losses([G_losses, D_losses], [\"gen\", \"disc\"], self.sample_dir + \"losses.png\")\n",
    "                    plot_losses([W_dist], [\"w_dist\"], self.sample_dir + \"dist.png\")\n",
    "                    plot_losses([grad_penalties],[\"grad_penalties\"], self.sample_dir + \"grad.png\")\n",
    "                    plot_losses([d_fake_losses, d_real_losses],[\"d_fake\", \"d_real\"], self.sample_dir + \"d_loss_components.png\")\n",
    "                i += 1\n",
    "\n",
    "\n",
    "    def sample(self, num_batches_sample, epoch):\n",
    "        decoded_seqs = []\n",
    "        for i in range(num_batches_sample):\n",
    "            z = to_var(torch.randn(self.batch_size, 128))\n",
    "            self.G.eval()\n",
    "            torch_seqs = self.G(z)\n",
    "            seqs = (torch_seqs.data).cpu().numpy()\n",
    "            decoded_seqs += [decode_one_seq(seq, self.inv_charmap) for seq in seqs]\n",
    "        self.G.train()\n",
    "        return decoded_seqs\n",
    "\n",
    "def main():\n",
    "    parser = argparse.ArgumentParser(description='FBGAN with Psipred Analyzer')\n",
    "    parser.add_argument(\"--run_name\", default=\"fbgan_psipred_demo\", help=\"Name for output files\")\n",
    "    parser.add_argument(\"--load_dir\", default=\"./checkpoint/realProt_50aa/\", help=\"Load existing data set\")\n",
    "    args = parser.parse_known_args()[0]\n",
    "    model = WGAN_LangGP(run_name=args.run_name)\n",
    "    model.train_model(args.load_dir)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "001fc968",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
