{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calc the frequences for samples generated in pre-train, trainning ,after trained\n",
    "#compare them with real samples\n",
    "#translate them into peptides\n",
    "from utils.utils import *\n",
    "import matplotlib.pyplot as plt\n",
    "import utils.language_helpers as lh\n",
    "from utils.bio_utils import *\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_table = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',\n",
    "}\n",
    "\n",
    "def countCorrectP(dna_seqs, verbose=False):\n",
    "    global codon_table\n",
    "    total = 0.0\n",
    "    correct = 0.0\n",
    "    for dna_seq in dna_seqs:\n",
    "        p_seq = \"\"\n",
    "        total += 1\n",
    "        if dna_seq[0:3] != 'ATG':\n",
    "            if verbose: print(\"Not valid gene (no ATG)\")\n",
    "            continue\n",
    "        for i in range(3, len(dna_seq), 3):\n",
    "            codon = dna_seq[i:i+3]\n",
    "            try:\n",
    "                aa = codon_table[codon]\n",
    "                p_seq += aa\n",
    "                if aa == '_': \n",
    "                    break\n",
    "            except:\n",
    "                if verbose: print(\"Error! Invalid Codon {} in {}\".format(codon, dna_seq))\n",
    "                break\n",
    "        if len(p_seq) <= 2: #needs to have stop codon and be of length greater than 2\n",
    "            if verbose: print(\"Error! Protein too short.\")\n",
    "        elif p_seq[-1] != '_':\n",
    "            if verbose: print(\"Error! No valid stop codon.\")\n",
    "        else:\n",
    "            correct+=1\n",
    "    return correct/total\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plo(aim_list,labellist,xlabel,ylim):\n",
    "    for i in range(len(labellist)):\n",
    "        plt.plot(aim_list[i],labellist[i])\n",
    "    plt.xlabel(xlabel=xlabel)\n",
    "    plt.ylim(bottom=ylim)\n",
    "    plt.legend(loc='lower right')\n",
    "\n",
    "def is_PInDNA(line):\n",
    "    for i in range(len(line)):\n",
    "        if i+1< len(line) and line[i]==\"P\" \\\n",
    "        and line[i+1].isalpha() and line[i+1]!='P':\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "def cutline(line):\n",
    "    mark = 0\n",
    "    for i in range(len(line)):\n",
    "        if line[i] == \"P\" or i == len(line) - 1 or line[i].isalpha()==False:\n",
    "            mark = i\n",
    "            break\n",
    "    return line[0:mark]\n",
    "    \n",
    "def divideline(lines):\n",
    "    dst = []\n",
    "    line = ''\n",
    "    for char in lines[0]:\n",
    "        line += char\n",
    "        if len(line) == 156:\n",
    "            dst.append(line)\n",
    "            line = ''\n",
    "    return dst\n",
    "\n",
    "def CalcFreq(path):\n",
    "    with open(path,encoding='utf-8') as f:\n",
    "        content = f.read()\n",
    "        lines = content.split('\\n')\n",
    "        # if len(lines) == 1:\n",
    "        #     lines = divideline(lines)\n",
    "        cnt_total = 0.0\n",
    "        cnt_PInDNA = 0.0\n",
    "        freq_A = 0.0\n",
    "        freq_T = 0.0\n",
    "        freq_G = 0.0\n",
    "        freq_C = 0.0\n",
    "        freq_P = 0.0\n",
    "        mean_len = 0.0\n",
    "        mean_correctP = countCorrectP(lines)\n",
    "        for line in lines:\n",
    "            if len(line) == 0:continue\n",
    "            cnt_total+=1\n",
    "            freq_P += (float(line.count('P'))/len(line))\n",
    "            if is_PInDNA(line):\n",
    "                cnt_PInDNA+=1\n",
    "            line  = cutline(line)\n",
    "            len_dna = len(line)\n",
    "            if len_dna == 0:\n",
    "                continue\n",
    "            mean_len += len_dna\n",
    "            freq_A += float(line.count('A'))/len_dna\n",
    "            freq_T += (float(line.count('T')))/len_dna\n",
    "            freq_G += (float(line.count('G')))/len_dna\n",
    "            freq_C += (float(line.count('C')))/len_dna\n",
    "        freq_A /= cnt_total\n",
    "        freq_T /= cnt_total\n",
    "        freq_G /= cnt_total\n",
    "        freq_C /= cnt_total\n",
    "        freq_P /= cnt_total\n",
    "        mean_len/= cnt_total\n",
    "        cnt_PInDNA/=cnt_total\n",
    "        plines = geneToProtein(lines,verbose=False)\n",
    "        mean_Lp = 0.0\n",
    "        for pline in plines:\n",
    "            mean_Lp+= len(pline)\n",
    "        if len(plines) == 0 :\n",
    "            mean_Lp = 0\n",
    "        else:\n",
    "            mean_Lp/= len(plines)\n",
    "        return freq_A,freq_T,freq_G,freq_C,freq_P,mean_len,cnt_PInDNA,mean_correctP,mean_Lp\n",
    "    \n",
    "def divide(path):\n",
    "    content = []\n",
    "    with open(path,encoding='utf-8') as f:\n",
    "        content = f.read()\n",
    "    lines = []\n",
    "    line = ''\n",
    "    # print(len(content))\n",
    "    # print(content[0])\n",
    "    for char in content:\n",
    "        line += char\n",
    "        if len(line) == 156:\n",
    "            lines.append(line)\n",
    "            line = ''\n",
    "    \n",
    "    with open(path,'w',encoding='utf-8') as f:\n",
    "        f.writelines([s + '\\n' for j, s in enumerate(lines)])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for i in range(12400):\n",
    "#     if i % 100 == 99:\n",
    "#         path = './samples/realProt_50aa/sampled_val_{}.txt'.format(i)\n",
    "#         divide(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_freq_A = []\n",
    "mean_freq_T = []\n",
    "mean_freq_G = []\n",
    "mean_freq_C = []\n",
    "mean_freq_P = []\n",
    "mean_Length = []\n",
    "mean_Lps = []\n",
    "cnt_PInS = []\n",
    "mean_correctPs = []\n",
    "for i in range(80):\n",
    "    # if i % 100 == 99:\n",
    "        # path = './samples/realProt_50aa/sampled_{}.txt'.format(i)\n",
    "        # path = './samples/realProt_50aa/sampled_{}.txt'.format(i)\n",
    "        path = './samples/realProt_50aa/sampled_{}.txt'.format(i)\n",
    "        freq_A,freq_T,freq_G,freq_C,freq_P,mean_len,cnt_PInDNA,mean_correctP,meanLp = CalcFreq(path)\n",
    "        mean_freq_A.append(freq_A)\n",
    "        mean_freq_T.append(freq_T)\n",
    "        mean_freq_G.append(freq_G)\n",
    "        mean_freq_C.append(freq_C)\n",
    "        mean_freq_P.append(freq_P)\n",
    "        mean_Length.append(mean_len)\n",
    "        cnt_PInS.append(cnt_PInDNA)\n",
    "        mean_correctPs.append(mean_correctP)\n",
    "        mean_Lps.append(meanLp)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_fb_freq_A = []\n",
    "mean_fb_freq_T = []\n",
    "mean_fb_freq_G = []\n",
    "mean_fb_freq_C = []\n",
    "mean_fb_freq_P = []\n",
    "mean_fb_Length = []\n",
    "mean_fb_correctPs = []\n",
    "fb_cnt_PInS = []\n",
    "mean_fb_Lps = []\n",
    "for i in range(150):\n",
    "        # path = './samples/fbgan_amp_demo/sampled_{}_preds.txt'.format(i+1)\n",
    "        path = './samples/fbgan_amp_demo/sample_val_{}.txt'.format(i+1)\n",
    "        freq_A,freq_T,freq_G,freq_C,freq_P,mean_len,cnt_PInDNA,mean_fb_correctP,meanLp = CalcFreq(path)\n",
    "        mean_fb_freq_A.append(freq_A)\n",
    "        mean_fb_freq_T.append(freq_T)\n",
    "        mean_fb_freq_G.append(freq_G)\n",
    "        mean_fb_freq_C.append(freq_C)\n",
    "        mean_fb_freq_P.append(freq_P)\n",
    "        mean_fb_Length.append(mean_len)\n",
    "        fb_cnt_PInS.append(cnt_PInDNA)\n",
    "        mean_fb_correctPs.append(mean_fb_correctP)\n",
    "        mean_fb_Lps.append(meanLp)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading dataset...\n",
      "('A', 'T', 'G', 'T', 'C', 'T', 'G', 'G', 'C', 'C', 'G', 'C', 'G', 'G', 'T', 'A', 'A', 'G', 'A', 'C', 'G', 'G', 'G', 'T', 'G', 'G', 'C', 'A', 'A', 'A', 'G', 'C', 'C', 'A', 'G', 'G', 'G', 'C', 'T', 'A', 'A', 'A', 'G', 'C', 'A', 'A', 'A', 'A', 'A', 'C', 'A', 'A', 'G', 'A', 'A', 'G', 'T', 'A', 'G', 'C', 'C', 'G', 'G', 'G', 'C', 'T', 'G', 'G', 'C', 'C', 'T', 'T', 'C', 'A', 'G', 'T', 'T', 'C', 'C', 'C', 'C', 'G', 'T', 'A', 'G', 'G', 'G', 'A', 'G', 'G', 'G', 'T', 'T', 'C', 'A', 'T', 'C', 'G', 'T', 'T', 'T', 'A', 'T', 'T', 'A', 'A', 'G', 'A', 'A', 'A', 'A', 'G', 'G', 'C', 'A', 'A', 'C', 'T', 'A', 'C', 'G', 'C', 'G', 'C', 'A', 'C', 'C', 'G', 'T', 'G', 'T', 'C', 'G', 'G', 'G', 'G', 'C', 'C', 'G', 'G', 'C', 'G', 'C', 'T', 'C', 'C', 'C', 'G', 'T', 'A', 'T', 'A', 'T', 'T', 'T', 'A')\n",
      "loaded 2048 lines in dataset\n"
     ]
    }
   ],
   "source": [
    "real_datas,_,__ = lh.load_dataset(max_length=156,max_n_examples=2048,data_dir='./data/AMP_dataset.fa')\n",
    "real_lines = []\n",
    "for data in real_datas:\n",
    "        line = ''\n",
    "        for char in data:\n",
    "                if char.isalpha():\n",
    "                        line += char\n",
    "        real_lines.append(line)\n",
    "# print(real_lines)\n",
    "store_path = './samples_real/realsamplesamp.txt'\n",
    "with open(store_path,mode='w+',encoding='utf-8') as f:\n",
    "        content = ''\n",
    "        for line in real_lines:\n",
    "                content+= line\n",
    "                content+='\\n'\n",
    "        f.write(content)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_real_freq_A = []\n",
    "mean_real_freq_T = []\n",
    "mean_real_freq_G = []\n",
    "mean_real_freq_C = []\n",
    "mean_real_freq_P = []\n",
    "mean_real_Length = []\n",
    "real_cnt_PInS = []\n",
    "mean_real_Lps = []\n",
    "\n",
    "freq_A,freq_T,freq_G,freq_C,freq_P,mean_len,cnt_PInDNA,_,meanLp = CalcFreq(store_path)\n",
    "mean_real_freq_A = [freq_A] * len(mean_fb_freq_A)\n",
    "mean_real_freq_T = [freq_T] * len(mean_fb_freq_A)\n",
    "mean_real_freq_G = [freq_G] * len(mean_fb_freq_A)\n",
    "mean_real_freq_C = [freq_C] * len(mean_fb_freq_A)\n",
    "mean_real_freq_P = [freq_P] * len(mean_fb_freq_A)\n",
    "mean_real_Length = [mean_len] * len(mean_fb_freq_A)\n",
    "real_cnt_PInS = [cnt_PInDNA] * len(mean_fb_freq_A)\n",
    "mean_real_Lps = [meanLp] * len(mean_fb_freq_A)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "mpl.rcParams.update({'font.size': 20})\n",
    "\n",
    "plt.figure(figsize=(20,10))\n",
    "plt.plot(mean_Length, label='mean len in pre-train')\n",
    "plt.plot(mean_fb_Length, label='mean len in feedback train')\n",
    "plt.plot(mean_real_Length,label='mean len of real samples')\n",
    "\n",
    "plt.title('Mean length of nucleotides chains per batch')\n",
    "plt.xlabel('Epoch')\n",
    "plt.ylim(bottom=0)\n",
    "plt.grid()\n",
    "\n",
    "plt.legend(loc='lower right')\n",
    "plt.savefig('./AMP/MeanLength.png')\n",
    "\n",
    "plt.figure(figsize=(20,10))\n",
    "plt.plot(mean_correctPs, label='freq of genes with correct structure in pre-train')\n",
    "plt.plot(mean_fb_correctPs, label='freq of genes with correct structure in feedback train')\n",
    "\n",
    "plt.title('frequence of genes with correct structure')\n",
    "plt.xlabel('Epoch')\n",
    "plt.ylim(bottom=0)\n",
    "plt.grid()\n",
    "\n",
    "plt.legend(loc='lower right')\n",
    "plt.savefig('./AMP/CorrectGene.png')\n",
    "\n",
    "plt.figure(figsize=(20,10))\n",
    "plt.plot(cnt_PInS, label='Freq of seqs in pre-train')\n",
    "plt.plot(fb_cnt_PInS, label='Freq of seqs in feedback train')\n",
    "plt.plot(real_cnt_PInS, label= 'Freq of real samples')\n",
    "plt.title('Frequences of sequences that has multiple nucleotides chains')\n",
    "plt.xlabel('Epoch')\n",
    "plt.ylim(bottom=0)\n",
    "plt.legend(loc='lower right')\n",
    "plt.grid()\n",
    "\n",
    "plt.savefig('./AMP/MultipleNC.png')\n",
    "\n",
    "plt.figure(figsize=(20,10))\n",
    "\n",
    "plt.plot(mean_Lps, label='Mean length of generated protein in Pre-train')\n",
    "plt.plot(mean_fb_Lps, label='Mean length of generated protein in feedback train')\n",
    "plt.plot(mean_real_Lps, label='Mean length of real protein')\n",
    "\n",
    "plt.title('Mean length of protein')\n",
    "plt.xlabel('Epoch')\n",
    "plt.ylim(bottom=0)\n",
    "plt.grid()\n",
    "\n",
    "plt.legend(loc='lower right')\n",
    "plt.savefig('./AMP/MeanLenP.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(20,13))\n",
    "\n",
    "plt.plot(mean_fb_freq_A, label='freq_A in feedback train',linewidth=3)\n",
    "plt.plot(mean_fb_freq_T, label='freq_T in feedback train',linewidth=3)\n",
    "plt.plot(mean_fb_freq_G, label='freq_G in feedback train',linewidth=3)\n",
    "plt.plot(mean_fb_freq_C, label='freq_C in feedback train',linewidth=3)\n",
    "plt.plot(mean_real_freq_A, label='freq_A of real samples',linewidth=3)\n",
    "plt.plot(mean_real_freq_T, label='freq_T of real samples',linewidth=3)\n",
    "plt.plot(mean_real_freq_G, label='freq_G of real samples',linewidth=3)\n",
    "plt.plot(mean_real_freq_C, label='freq_C of real samples',linewidth=3)\n",
    "\n",
    "plt.xlabel('Epoch')\n",
    "plt.ylim(bottom=0)\n",
    "plt.grid()\n",
    "plt.title('Percentage of nucleotides per chain in feedback train')\n",
    "plt.legend(loc='lower left')\n",
    "\n",
    "plt.savefig('./AMP/NDistribution_fb.png')\n",
    "\n",
    "plt.figure(figsize=(20,13))\n",
    "plt.plot(mean_freq_A, label='freq_A in pre-train',linewidth=2)\n",
    "plt.plot(mean_freq_T, label='freq_T in pre-train',linewidth=2)\n",
    "plt.plot(mean_freq_G, label='freq_G in pre-train',linewidth=2)\n",
    "plt.plot(mean_freq_C, label='freq_C in pre-train',linewidth=2)\n",
    "plt.plot(mean_real_freq_A[0:len(mean_freq_A)], label='freq_A of real samples',linewidth=2)\n",
    "plt.plot(mean_real_freq_T[0:len(mean_freq_A)], label='freq_T of real samples',linewidth=2)\n",
    "plt.plot(mean_real_freq_G[0:len(mean_freq_A)], label='freq_G of real samples',linewidth=2)\n",
    "plt.plot(mean_real_freq_C[0:len(mean_freq_A)], label='freq_C of real samples',linewidth=2)\n",
    "plt.grid()\n",
    "\n",
    "plt.xlabel('Epoch')\n",
    "plt.ylim(bottom=0)\n",
    "plt.title('Percentage of nucleotides per chain in pre-train')\n",
    "plt.legend(loc='upper right')\n",
    "plt.savefig('./AMP/NDistribution_pre.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mean_test_freq_A = []\n",
    "# mean_test_freq_T = []\n",
    "# mean_test_freq_G = []\n",
    "# mean_test_freq_C = []\n",
    "# mean_test_freq_P = []\n",
    "# mean_test_Length = []\n",
    "# test_cnt_PInS = []\n",
    "# test_path = './samples/fbgan_amp_demo/test.txt'\n",
    "# freq_A,freq_T,freq_G,freq_C,freq_P,mean_len,cnt_PInDNA = CalcFreq(store_path)\n",
    "# mean_test_freq_A = [freq_A] * len(mean_fb_freq_A)\n",
    "# mean_test_freq_T = [freq_T] * len(mean_fb_freq_A)\n",
    "# mean_test_freq_G = [freq_G] * len(mean_fb_freq_A)\n",
    "# mean_test_freq_C = [freq_C] * len(mean_fb_freq_A)\n",
    "# mean_test_freq_P = [freq_P] * len(mean_fb_freq_A)\n",
    "# mean_test_Length = [mean_len] * len(mean_fb_freq_A)\n",
    "# test_cnt_PInS = [cnt_PInDNA] * len(mean_fb_freq_A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plt.figure(figsize=(20,13))\n",
    "\n",
    "# plt.plot(mean_fb_freq_A, label='freq_A in feedback train',linewidth=3)\n",
    "# plt.plot(mean_fb_freq_T, label='freq_T in feedback train',linewidth=3)\n",
    "# plt.plot(mean_fb_freq_G, label='freq_G in feedback train',linewidth=3)\n",
    "# plt.plot(mean_fb_freq_C, label='freq_C in feedback train',linewidth=3)\n",
    "# plt.plot(mean_test_freq_A, label='freq_A of test samples',linewidth=3)\n",
    "# plt.plot(mean_test_freq_T, label='freq_T of test samples',linewidth=3)\n",
    "# plt.plot(mean_test_freq_G, label='freq_G of test samples',linewidth=3)\n",
    "# plt.plot(mean_test_freq_C, label='freq_C of test samples',linewidth=3)\n",
    "# plt.xlabel('Epoch')\n",
    "# plt.ylim(bottom=0)\n",
    "# plt.title('Percentage of nucleotides per chain in feedback train')\n",
    "# plt.legend(loc='lower right')\n",
    "# plt.savefig('./AMP/NDistribution_test.png')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('swb')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "60bff9b272fedff1fa90ce0dbeee5551a5297cee346565f639bed6524c8875b2"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
